compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs

```diff
@@ -110,15 +124,48 @@ impl Context {
 
         Self::replace_result_ids(dfg, &old_results, &new_results);
 
-        Self::cache_instruction(instruction, new_results, dfg, instruction_result_cache);
+        // If we just inserted an `Instruction::EnableSideEffects`, we need to update `side_effects_enabled_var`
+        // so that we use the correct set of constrained values in future.
+        if let Instruction::EnableSideEffects { condition } = instruction {
+            *side_effects_enabled_var = condition;
+        };
+
+        Self::cache_instruction(
+            instruction,
+            new_results,
+            dfg,
+            instruction_result_cache,
+            constrained_values.entry(*side_effects_enabled_var).or_default(),
+        );
```

Comment:
```suggestion
        Self::cache_instruction(
            instruction,
            new_results,
            dfg,
            instruction_result_cache,
            constrained_values.entry(*side_effects_enabled_var).or_default(),
        );
        
         // If we just inserted an `Instruction::EnableSideEffects`, we need to update `side_effects_enabled_var`
        // so that we use the correct set of constrained values in future.
        if let Instruction::EnableSideEffects { condition } = instruction {
            *side_effects_enabled_var = condition;
        };
```

Swapping these around to avoid misunderstandings about caching after switching the `side_effects_enabled_var`. This doesn't change behaviour as `Instruction::EnableSideEffects` doesn't contribute to any caches.